EKF/SensorRangeFinder.hpp

```diff
@@ -0,0 +1,150 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2020 Estimation and Control Library (ECL). All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name ECL nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file SensorRangeFinder.hpp
+ * Range finder class containing all the required checks
+ *
+ * @author Mathieu Bresciani <brescianimathieu@gmail.com>
+ *
+ */
+#pragma once
+
+#include "Sensor.hpp"
+#include <matrix/math.hpp>
+
+namespace estimator
+{
+namespace sensor
+{
+
+class SensorRangeFinder : public Sensor
+{
+public:
+	SensorRangeFinder() = default;
+	~SensorRangeFinder() override = default;
+
+	void runChecks(uint64_t current_time_us, const matrix::Dcmf &R_to_earth);
+	bool isHealthy() const override { return _is_sample_valid; }
+	bool isDataHealthy() const override { return _is_sample_ready && _is_sample_valid; }
+
+	void setSample(rangeSample sample) {
```

Comment:
IMO this class should contain the value and not be dependent on the scope of the referenced value. For example, let's imagine that we have a class member `rangeSample _rng` and that in some function we do:
```c++
void createNewSample() {
   rangeSample new_sample{10.f, 2e6, 100};
   _rng.setSample(new_sample);
}
```
Since the original variable is destroyed at the end of the function, the reference will be broken.

Reply:
Yeah, but that is why you have am private member `_range_sample` in your `SensorRangeFinder`class. Inside setSample(), you make a copy assignment of the sample to the private member. Since the sample does not contain any pointers, after the assignment the private member `_range_sample` is completely independent of the outside sample.
But passing the argument to the function by const-ref avoids the construction of an intermediate sample object. This intermediate sample would only get assigned to the private member variable before getting destroyed again. Instead we could assign the outside sample that got passed by reference.

Reply:
Of course, yes, I think I was tired xD